File: src/main/java/net/azisaba/townia/manager/MessageManager.java
package net.azisaba.townia.manager;

import net.azisaba.townia.Townia;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
import org.bukkit.command.CommandSender;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.entity.Player;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.logging.Level;

public class MessageManager {

    private final Townia plugin;
    private final Map<String, Map<String, Object>> allMessages = new HashMap<>();
    private String defaultLang;
    private final MiniMessage miniMessage = MiniMessage.miniMessage();

    public MessageManager(Townia plugin) {
        this.plugin = plugin;
        loadAllMessages();
    }

    public void loadAllMessages() {
        allMessages.clear();
        defaultLang = plugin.getConfig().getString("default-language", "ja");

        File messagesDir = new File(plugin.getDataFolder(), "messages");
        if (!messagesDir.exists()) messagesDir.mkdirs();

        for (String code : List.of("en", "ja")) {
            File target = new File(messagesDir, "lang_" + code + ".yml");
            if (!target.exists()) {
                try (InputStream in = plugin.getResource("messages/lang_" + code + ".yml")) {
                    if (in != null) {
                        try (OutputStream out = new FileOutputStream(target)) {
                            in.transferTo(out);
                        }
                    }
                } catch (IOException e) {
                    plugin.getLogger().log(Level.WARNING, "Failed to extract lang_" + code + ".yml", e);
                }
            }
        }

        File[] langFiles = messagesDir.listFiles((dir, name) -> name.matches("lang_\\w+\\.yml"));
        if (langFiles == null) return;

        for (File langFile : langFiles) {
            String langCode = langFile.getName().replace("lang_", "").replace(".yml", "");
            try (InputStreamReader reader = new InputStreamReader(new FileInputStream(langFile), StandardCharsets.UTF_8)) {
                FileConfiguration cfg = YamlConfiguration.loadConfiguration(reader);
                Map<String, Object> messages = new LinkedHashMap<>();
                for (String key : cfg.getKeys(true)) {
                    if (!cfg.isConfigurationSection(key)) {
                        messages.put(key, cfg.get(key));
                    }
                }
                allMessages.put(langCode, messages);
                plugin.getLogger().info("Loaded language file: " + langFile.getName() + " (" + messages.size() + " keys)");
            } catch (IOException e) {
                plugin.getLogger().log(Level.WARNING, "Failed to load " + langFile.getName(), e);
            }
        }
    }

    public void sendMessage(CommandSender sender, String key, String... replacements) {
        String lang = getLang(sender);
        Object messageObject = getMessageObject(lang, key);

        List<?> lines;
        if (messageObject instanceof List<?> l) {
            lines = l;
        } else {
            lines = Collections.singletonList(messageObject);
        }

        String prefix = plugin.getTowniaConfig().getPrefix();
        for (Object line : lines) {
            String raw = prefix + replacePlaceholders(String.valueOf(line), replacements);
            sender.sendMessage(miniMessage.deserialize(raw));
        }
    }

    public void sendActionBar(Player player, String key, String... replacements) {
        String lang = getLang(player);
        Object messageObject = getMessageObject(lang, key);

        String raw;
        if (messageObject instanceof List<?> l && !l.isEmpty()) {
            raw = String.valueOf(l.get(0));
        } else {
            raw = String.valueOf(messageObject);
        }

        raw = replacePlaceholders(raw, replacements);
        player.sendActionBar(miniMessage.deserialize(raw));
    }

    public String getPlainMessage(CommandSender sender, String key, String... replacements) {
        String lang = getLang(sender);
        Object obj = getMessageObject(lang, key);
        String raw;
        if (obj instanceof List<?> list && !list.isEmpty()) {
            raw = String.valueOf(list.getFirst());
        } else {
            raw = String.valueOf(obj);
        }
        raw = replacePlaceholders(raw, replacements);
        Component component = miniMessage.deserialize(raw);
        return PlainTextComponentSerializer.plainText().serialize(component);
    }

    private String getLang(CommandSender sender) {
        if (sender instanceof Player player) {
            String clientLang = player.locale().getLanguage().toLowerCase();
            if (allMessages.containsKey(clientLang)) return clientLang;
        }
        return defaultLang;
    }

    private Object getMessageObject(String lang, String key) {
        Map<String, Object> messages = allMessages.getOrDefault(lang, allMessages.get(defaultLang));
        if (messages == null) return "<red>No language file loaded for: " + lang;
        return messages.getOrDefault(key, "<red>Missing message key: " + key);
    }

    private String replacePlaceholders(String text, String... replacements) {
        if (replacements.length % 2 != 0) {
            throw new IllegalArgumentException("Replacements must be in pairs (key, value, key, value, …)");
        }
        for (int i = 0; i < replacements.length; i += 2) {
            String key = replacements[i];
            if (!key.startsWith("{") && !key.endsWith("}")) {
                key = "{" + key + "}";
            }
            text = text.replace(key, replacements[i + 1]);
        }
        return text;
    }
}

File: src/main/java/net/azisaba/townia/manager/MessageManager.kt
package net.azisaba.townia.manager

import net.azisaba.townia.Townia
import net.kyori.adventure.text.minimessage.MiniMessage
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer
import org.bukkit.command.CommandSender
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.Player
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.util.logging.Level

class MessageManager(private val plugin: Townia) {

    private val allMessages = HashMap<String, Map<String, Any>>()
    private lateinit var defaultLanguage: String
    val miniMessage: MiniMessage = MiniMessage.miniMessage()

    init {
        loadAllMessages()
    }

    fun loadAllMessages() {
        allMessages.clear()
        defaultLanguage = plugin.config.getString("default-language", "ja") ?: "ja"

        val messagesDir = File(plugin.dataFolder, "messages")
        if (!messagesDir.exists()) messagesDir.mkdirs()

        for (code in listOf("ja", "en")) {
            val targetFile = File(messagesDir, "lang_$code.yml")
            if (!targetFile.exists()) {
                try {
                    plugin.getResource("messages/lang_$code.yml")?.use { inputStream ->
                        targetFile.outputStream().use { outputStream ->
                            inputStream.copyTo(outputStream)
                        }
                    }
                } catch (e: IOException) {
                    plugin.logger.log(Level.WARNING, "Failed to extract lang_$code.yml", e)
                }
            }
        }

        val langFiles = messagesDir.listFiles { _, name -> name.matches(Regex("lang_\\w+\\.yml")) }
        if (langFiles == null) return

        for (langFile in langFiles) {
            val langCode = langFile.name.replace("lang_", "").replace(".yml", "")

            val messages = LinkedHashMap<String, Any>()
            try {
                InputStreamReader(FileInputStream(langFile), StandardCharsets.UTF_8).use { reader ->
                    val cfg = YamlConfiguration.loadConfiguration(reader)
                    for (key in cfg.getKeys(true)) {
                        if (!cfg.isConfigurationSection(key)) {
                            cfg.get(key)?.let { messages.put(key, it) }
                        }
                    }
                }
            } catch (e: IOException) {
                plugin.logger.log(Level.WARNING, "Failed to load ${langFile.name}", e)
            }

            try {
                plugin.getResource("messages/lang_$langCode.yml")?.use { input ->
                    InputStreamReader(input, StandardCharsets.UTF_8).use { reader ->
                        val internalCfg = YamlConfiguration.loadConfiguration(reader)
                        val internalVersion = internalCfg.getInt("version", 1)
                        val currentVersion = (messages["version"] as? Int) ?: 0

                        if (currentVersion < internalVersion) {
                            plugin.logger.info("Outdated lang_$langCode.yml detected (v$currentVersion < v$internalVersion). Updating...")
                            val bakFile = File(messagesDir, "lang_$langCode.yml.v$currentVersion.bak")
                            langFile.renameTo(bakFile)

                            // Re-save the default resource
                            plugin.getResource("messages/lang_$langCode.yml")?.use { inputStream ->
                                langFile.outputStream().use { outputStream ->
                                    inputStream.copyTo(outputStream)
                                }
                            }
                            
                            messages.clear()
                            InputStreamReader(FileInputStream(langFile), StandardCharsets.UTF_8).use { freshReader ->
                                val freshCfg = YamlConfiguration.loadConfiguration(freshReader)
                                for (key in freshCfg.getKeys(true)) {
                                    if (!freshCfg.isConfigurationSection(key)) {
                                        freshCfg.get(key)?.let { messages[key] = it }
                                    }
                                }
                            }
                        } else {
                            for (key in internalCfg.getKeys(true)) {
                                if (!internalCfg.isConfigurationSection(key) && !messages.containsKey(key)) {
                                    internalCfg.get(key)?.let { messages.put(key, it) }
                                }
                            }
                        }
                    }
                }
            } catch (e: IOException) {
                plugin.logger.log(Level.WARNING, "Failed to load internal lang_$langCode.yml", e)
            }

            allMessages[langCode] = messages
            plugin.logger.info("Loaded language file: ${langFile.name} (${messages.size} keys)")
        }
    }

    fun sendMessage(sender: CommandSender, key: String, vararg replacements: String) {
        sendMessageInternal(sender, key, true, *replacements)
    }

    fun sendMessageWithoutPrefix(sender: CommandSender, key: String, vararg replacements: String) {
        sendMessageInternal(sender, key, false, *replacements)
    }

    private fun sendMessageInternal(sender: CommandSender, key: String, withPrefix: Boolean, vararg replacements: String) {
        val language = getLanguage(sender)
        val lines: List<*> = when (val messageObject = getMessageObject(language, key)) {
            is List<*> -> messageObject
            else -> listOf(messageObject)
        }

        val prefix = if (withPrefix) plugin.towniaConfig.prefix else ""
        for (line in lines) {
            val raw = replaceLegacyColors(prefix + replacePlaceholders(line.toString(), *replacements))
            sender.sendMessage(miniMessage.deserialize(raw))
        }
    }

    fun sendActionBar(player: Player, key: String, vararg replacements: String) {
        val language = getLanguage(player)
        val messageObject = getMessageObject(language, key)

        val raw = if (messageObject is List<*> && messageObject.isNotEmpty()) {
            messageObject.first().toString()
        } else {
            messageObject.toString()
        }

        val processedRaw = replaceLegacyColors(replacePlaceholders(raw, *replacements))
        player.sendActionBar(miniMessage.deserialize(processedRaw))
    }

    fun sendTitle(player: Player, mainKey: String, subKey: String, vararg replacements: String) {
        val language = getLanguage(player)
        val mainObj = getMessageObject(language, mainKey)
        val subObj = getMessageObject(language, subKey)

        val mainRaw = if (mainObj is List<*> && mainObj.isNotEmpty()) mainObj.first().toString() else mainObj.toString()
        val subRaw = if (subObj is List<*> && subObj.isNotEmpty()) subObj.first().toString() else subObj.toString()

        val processedMain = replaceLegacyColors(replacePlaceholders(mainRaw, *replacements))
        val processedSub = replaceLegacyColors(replacePlaceholders(subRaw, *replacements))

        val title = net.kyori.adventure.title.Title.title(
            miniMessage.deserialize(processedMain),
            miniMessage.deserialize(processedSub),
            net.kyori.adventure.title.Title.Times.times(
                java.time.Duration.ofMillis(500),
                java.time.Duration.ofMillis(3000),
                java.time.Duration.ofMillis(1000)
            )
        )
        player.showTitle(title)
    }

    fun getPlainMessage(sender: CommandSender, key: String, vararg replacements: String): String {
        val language = getLanguage(sender)
        val obj = getMessageObject(language, key)

        val raw = if (obj is List<*> && obj.isNotEmpty()) {
            obj.first().toString()
        } else {
            obj.toString()
        }

        val processedRaw = replaceLegacyColors(replacePlaceholders(raw, *replacements))
        val component = miniMessage.deserialize(processedRaw)
        return PlainTextComponentSerializer.plainText().serialize(component)
    }

    fun getRawMessage(sender: CommandSender, key: String, vararg replacements: String): String {
        val language = getLanguage(sender)
        val obj = getMessageObject(language, key)

        val raw = if (obj is List<*> && obj.isNotEmpty()) {
            obj.first().toString()
        } else {
            obj.toString()
        }

        return replaceLegacyColors(replacePlaceholders(raw, *replacements))
    }


    private fun getLanguage(sender: CommandSender): String {
        if (sender is Player) {
            val clientLanguage = sender.locale().language.lowercase()
            if (allMessages.containsKey(clientLanguage)) return clientLanguage
        }
        return defaultLanguage
    }

    private fun getMessageObject(language: String, key: String): Any? {
        val messages = allMessages[language]
        val defaultMessages = allMessages[defaultLanguage]

        if (messages != null && messages.containsKey(key)) {
            return messages[key] ?: "<red>Missing message key: $key"
        }

        if (defaultMessages != null && defaultMessages.containsKey(key)) {
            return defaultMessages[key] ?: "<red>Missing message key: $key"
        }

        return "<red>Missing message key: $key"
    }

    private fun replacePlaceholders(text: String, vararg replacements: String): String {
        require(replacements.size % 2 == 0) { "Replacements must be in pairs (key, value, key, value, ...)" }
        var result = text
        replacements.asSequence().chunked(2).forEach { pair ->
            if (pair.size == 2) {
                val rawKey = pair[0]
                val value = pair[1]
                val formattedKey = if (rawKey.startsWith("{") && rawKey.endsWith("}")) rawKey else "{${rawKey}}"
                result = result.replace(formattedKey, value)
            }
        }
        return result
    }

    private fun replaceLegacyColors(text: String): String {
        return text.replace(Regex("[&§]([0-9a-fk-orA-FK-OR])")) { match ->
            when (match.groupValues[1].lowercase()) {
                "0" -> "<black>"
                "1" -> "<dark_blue>"
                "2" -> "<dark_green>"
                "3" -> "<dark_aqua>"
                "4" -> "<dark_red>"
                "5" -> "<dark_purple>"
                "6" -> "<gold>"
                "7" -> "<gray>"
                "8" -> "<dark_gray>"
                "9" -> "<blue>"
                "a" -> "<green>"
                "b" -> "<aqua>"
                "c" -> "<red>"
                "d" -> "<light_purple>"
                "e" -> "<yellow>"
                "f" -> "<white>"
                "k" -> "<obfuscated>"
                "l" -> "<bold>"
                "m" -> "<strikethrough>"
                "n" -> "<underlined>"
                "o" -> "<italic>"
                "r" -> "<reset>"
                else -> match.value
            }
        }
    }
}
